Use a set to ensure we don't double count peers

ptschip opened the pull request:
We can receive multiple VERSION or SENDCMPCT messages which can
cause us to double count those peer types. By using a set
we can easily filter out the duplicates.

ptschip committed:
Use a set to ensure we don't double count peers

We can receive multiple VERSION or SENDCMPCT messages which can
cause us to double count those peer types. By using a set
we can easily filter out the duplicates.

gandrewstone commented:
If we ever accidentally don't erase from the set upon disconnect, then a node will be able to consume all of our memory by repeatedly triggering the problem.  It seems to me that these sets can never exceed the total number of connected peers.  I would like to see a DbgAssert before the set insert that checks if the set size > # peers, asserts in debug mode and skips the insertion in release mode.  By skipping the insertion, yes the counts will be screwed up (but they are screwed up anyway) but the sets won't grow uncontrollably.

ptschip committed:
Release the node reference

When we're done setting up the request of a transaction we can
finally release the ref and delete the node from the items list.

ptschip commented:
done

On 26/06/2019 6:23 p.m., Andrew Stone wrote:
>
> If we ever accidentally don't erase from the set upon disconnect, then
> a node will be able to consume all of our memory by repeatedly
> triggering the problem. It seems to me that these sets can never
> exceed the total number of connected peers. I would like to see a
> DbgAssert before the set insert that checks if the set size > # peers,
> asserts in debug mode and skips the insertion in release mode. By
> skipping the insertion, yes the counts will be screwed up (but they
> are screwed up anyway) but the sets won't grow uncontrollably.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1821?email_source=notifications&email_token=ADCELUH2DCKW6JD3EVSW4EDP4QJCRA5CNFSM4H3U4R3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYVH7IY#issuecomment-506101667>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADCELUEVBUWJJWGJY7JJDSDP4QJCRANCNFSM4H3U4R3A>.
>

ptschip committed:
Do not allow the sets for counting thin peers to grow unbounded

If somehow the sets get larger than the number of nodes connected
then assert (if in debug) and do nothing and return if running
a non debug build.
